src/Lean/Elab/Parallel.lean

```diff
@@ -0,0 +1,736 @@
+/-
+Copyright (c) 2025 Lean FRO, LLC. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Kim Morrison
+-/
+module
+
+prelude
+public import Lean.Elab.Task
+import Init.System.IO
+
+/-!
+# Iterator-based parallelization for Lean's tactic monads.
+
+This file provides utilities for running computations in parallel using Lean's task system,
+with support for collecting results in different ways.
+
+## Main functions
+
+For each monad (`IO`, `CoreM`, `MetaM`, `TermElabM`, `TacticM`), the following functions are provided:
+
+- `par` / `par'`
+  - Run jobs in parallel, collect results in original order
+  - Takes `List (MonadM α)`, returns `MonadM (List (Except Error (α × State)))` / `MonadM (List (Except Error α))`
+  - All tasks run in parallel, results returned in input order
+  - `par` returns state information, `par'` discards state
+  - Final state is restored to initial state (before tasks ran)
+  - Errors wrapped in `Except` so all results are collected
+
+- `parIter` / `parIterWithCancel`
+  - Run jobs in parallel, iterate over results in original order
+  - Takes `List (MonadM α)`, returns iterator
+  - `parIterWithCancel` also returns cancellation hook
+
+- `parIterGreedy` / `parIterGreedyWithCancel`
+  - Run jobs in parallel, iterate over results in completion order (greedily)
+  - Takes `List (MonadM α)`, returns iterator
+  - `parIterGreedyWithCancel` also returns cancellation hook
+
+- `parFirst`
+  - Run jobs in parallel, return first successful result (by completion order)
+  - Cancels remaining tasks after first success (by default)
+  - Throws error if all tasks fail
+
+## Implementation notes
+
+The greedy iterator-based functions use `IO.waitAny'` internally to wait for task completion in any order.
+The ordered iterator-based functions process tasks sequentially in the original order.
+
+**State threading in iterators:**
+The iterators (`parIter`, `parIterGreedy`, and their `WithCancel` variants) preserve state from each
+completed task. When you map over an iterator with a monadic function, the monad state will be that at
+the conclusion of the monadic action that produced each value. This means:
+- For `parIter`: State is threaded sequentially in the original task order
+- For `parIterGreedy`: State is threaded in task completion order
+
+This allows you to observe state changes (like logged messages, modified metavariable contexts, etc.)
+as tasks complete, unlike `par`/`par'` which restore the initial state after collecting all results.
+
+Iterators do not have `Finite` instances, as we cannot prove termination from the available
+information. For consumers that require `Finite` (like `.toList`), use `.allowNontermination.toList`.
+-/
+
+@[expose] public section
```

Comment:
Should it really be exposed? Could we avoid either in new files?

Reply:
Removed.